### PR TITLE
add 'lean' mode for Lean Theorem Prover files

### PR DESCRIPTION
**Pull request by soonhokong**
We're using it at https://leanprover.github.io/tutorial

### CODE REVIEW

File: lib/ace/mode/lean_highlight_rules.js [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+                regex : "[[({]"
+            }, {
+                token : "paren.rparen",
+                regex : "[\\])}]"
+            }, {
+                token : "text",
+                regex : "\\s+"
+            }
+        ],
+        "comment" : [
```
Comment (nightwing) [inline lib/ace/mode/lean_highlight_rules.js]: This is better written as `[ {token: "comment", regex: "-/", next: "start"}, {defaultToken: "comment"} ]`
Comment spanning whole line trick was necessary in old versions where defaultToken wasn't supported.

File: lib/ace/mode/lean_highlight_rules.js [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+            }
+        ],
+        "qstring" : [
+            {
+                token : "string",
+                regex : "(?:(?:\\\\.)|(?:[^'\\\\]))*?'",
+                next : "start"
+            }, {
+                token : "string",
+                regex : '.+'
```
Comment (nightwing) [inline lib/ace/mode/lean_highlight_rules.js]: same here

File: lib/ace/ext/modelist.js [hunk 3]
```diff
@@ -99,6 +99,7 @@ var supportedModes = {
     JSX:         ["jsx"],
     Julia:       ["jl"],
     LaTeX:       ["tex|latex|ltx|bib"],
+    Lean:        ["lean|hlean"],
```
Comment (nightwing) [inline lib/ace/ext/modelist.js]: Could you also add a small demo lean file to https://github.com/ajaxorg/ace/tree/master/demo/kitchen-sink/docs, just  a few lines would be fine

**Comment by soonhokong** [comment]
@nightwing, thanks for the comments. I'll fix them.

**Comment by soonhokong** [comment]
@nightwing, I've added things that you suggested. Could you review those commits? Thank you!

File: demo/kitchen-sink/docs/lean.lean [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+                ... = |a| * -b  : {(abs_of_nonpos H1)⁻¹}
+                ... = |a| * |b| : {(abs_of_nonpos H2)⁻¹}))
+
+  theorem abs_mul_self (a : A) : |a| * |a| = a * a :=
+  abs.by_cases rfl !neg_mul_neg
+end
+
+/- TODO: Multiplication and one, starting with mult_right_le_one_le. -/
+
+end algebra
```
Comment (nightwing) [inline demo/kitchen-sink/docs/lean.lean]: Could you please make this a _few lines_. Something closer to 10 than 500:)

**Comment by soonhokong** [comment]
@nightwing, I just made a destructive update https://github.com/ajaxorg/ace/commit/8f49d3d89784095089bf0885268b4e87e2fbc69c. It's 10 lines long.

File: lib/ace/mode/lean_highlight_rules.js [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+                token : "paren.rparen",
+                regex : "[\\])}]"
+            }, {
+                token : "text",
+                regex : "\\s+"
+            }
+        ],
+        "comment" : [ {token: "comment", regex: "-/", next: "start"},
+                      {defaultToken: "comment"} ],
+        "qqstring" : [ {token : "string", regex : '(?:(?:\\\\.)|(?:[^"\\\\]))*?"', next : "start" },
```
Comment (nightwing) [inline lib/ace/mode/lean_highlight_rules.js]: This should be something like

```js
[ {
     token : "constant.language.escape",
     regex : /\\./
}, {
     token : "string",
     regex : '"',
     next : "start" 
}, {
    defaultToken: "string"
}]
```

This depends on whether strings are multiline or not, and are there special escape patterns like \uXXXX

**Comment by soonhokong** [comment]
@nightwing, thanks for the comment. I've cleaned up many things and updated a rule for string. Please review them, esp 34d80e81cafe90ea78bdb5fcc05fad8a0076bbba.

**Comment by soonhokong** [comment]
@nightwing, FYI, 706ee70affc739b5e6943e8745b2d09b1f6352b9 is pushed.

**Comment by nightwing** [comment]
Pulled. 
Thank you!

**Comment by soonhokong** [comment]
@nightwing, really appreciate your help!

**Pull request merged**